### ISSUE TITLE
Add a way to specify how to evaluate models on a grid

### ISSUE THREAD

**Issue by hamogu** [astropy/astropy#4572]
Currently models are evaluated at a set of x values. However, in many situations we really need to integrate a model in the range [lower bin edge, upper bin edge]. For example, when fitting a PSF to an image, we want the total flux that the PSF predicts in a certain pixel (integrated over the pixel), not just the flux in the pixel center position.

@larrybradley raised this here: https://github.com/astropy/photutils/issues/48#issuecomment-43771581

**Comment by eteq** [comment]
I wonder if this should be integrated with the `render` functionality new in 1.1 (http://docs.astropy.org/en/stable/api/astropy.modeling.Model.html#astropy.modeling.Model.render and http://docs.astropy.org/en/stable/modeling/bounding-boxes.html#bounding-boxes)?

Also, I think this requires a "pre-step" that might also be useful: defining an interface to integrate a model over given bounds.  Ideally the base `Model` class would implement this numerically (presumably with `scipy.integrate.quad`/`scipy.integrate.dblquad`), but specific models can provide optimized versions when available.

**Comment by eteq** [comment]
Oh, and there's another way this might work: design a set of "container" models that wrap existing models to do the discretization/integration.  Not sure if that's better/worse, but a thought.

**Comment by pllim** [comment]
Like I mentioned in a previous meeting, some thoughts need to go into how flux is going to be supported as `Quantity`, when it comes time to support `Quantity` in `modeling`.

**Comment by eteq** [comment]
Definitely, @pllim.  Hopefully that's mostly orthogonal to these items, though, as that will have to wait on full `Quantity` support...

**Comment by eteq** [comment]
@hamogu et al.  This is now explicitly in photutils as part of astropy/photutils#159 - you can use the `PRFAdapter` to wrap a model in this manner.

So I wonder if we should close this issue?  I guess the big question is whether or not this kind of functionality is generically useful, or mostly confined to PSF/PRF modeling?

**Comment by larrybradley** [comment]
@eteq  I think this functionality is useful for any kind of profile fitting (e.g. fitting Sersic profiles to extended sources).  Note that `astropy.convolution.utils.discretize_model` can evaluate models on a grid, but that cannot be used within the fitting framework.  I think this issue is worth keeping open.  I may work on it myself at some point.

**Comment by astrofrog** [comment]
I think we definitely want something in the core package, not just in the photometry package though?

**Comment by astrofrog** [comment]
@hamogu - in this issue, are you also interested in **fitting** discrete models or just evaluating them? I was discussing with @gogrean at SciPy 2016 that it would be nice to be able to discretize models on-the-fly while fitting. A wrapper class would probably make sense for this. Let's say we had a wrapper class for models that made a model behave like a discrete model, one could then have a method on Model classes that is e.g.:

```python
dmodel = model.as_discrete_model(mode='subpixel', subpixel=10, bounds=...)
```

This could then be used for fitting, evaluation, etc? (and the discrete model would take integers for the values at which to be evaluated?)

**Comment by pllim** [comment]
@nden has a PR open for `TabularModel`, which I this is relevant? See #5105 .

**Comment by hamogu** [comment]
It's for fitting, not just for simply evaluating it (but obviously evaluating is a prerequisite for fitting). Apart from PDF fitting this comes up in every fit to an x-ray spectrum. Bins in CCD sources are typically so wide that they might contain several emission lines. Thus, it's not enough to evaluate the model at the midpoint, instead one has to really integrate the model over the full bin. As an added complication, the bins are typically of a different bin width, I.e. the data is dynamically binned to reach a certain number of counts in each bin.

**Comment by saimn** [comment]
I tried to use `astropy.convolution.utils.discretize_model` currently just to evaluate a model on a grid with oversampling and integral (and replace custom code), but it doesn't work because it uses `np.meshgrid` in "cartesian mode" (without `indexing='ij'`) which results in a flipped array compared to the usual way to evaluate a model (with `np.mgrid`) :cry: 
Then I would also been interested in fitting with an oversampled / integrated model. 
Photutils's `PRFAdapter` is a first step but only allows integrated model I think, which is slow. It would be good to have a more general way to do this directly in Astropy.
